driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -0,0 +1,760 @@
+     * @param value The value the codec should accept; must not be {@code null}.
+     * @param fqdn The fully-qualified name of the table column or UDT field to which the overriding codec applies;
+     * if {@code null}, no overriding codec will be located and the globally-registered codec will be returned instead.
+     * @return An overriding codec for the given keyspace, table and column, if any, otherwise, a globally-registered suitable codec.
+     * @throws CodecNotFoundException if an overriding codec is found but does not accept the given value;
+     * or if no suitable codec could be found at all.
+     */
+    public <T> TypeCodec<T> codecFor(DataType cqlType, T value, String fqdn) {
+        checkNotNull(cqlType, "Parameter cqlType cannot be null");
+        checkNotNull(value, "Parameter value cannot be null");
+        TypeCodec<T> codec = findOverridingCodec(cqlType, value, fqdn);
+        if (codec != null)
+            return codec;
+        return codecFor(cqlType, value);
+    }
+
+    private <T> TypeCodec<T> lookupCache(DataType cqlType, TypeToken<T> javaType) {
+        logger.debug("Looking up codec for {} <-> {}", cqlType, javaType);
+        CacheKey cacheKey = new CacheKey(javaType, cqlType);
+        try {
+            return (TypeCodec<T>)cache.get(cacheKey);
+        } catch (UncheckedExecutionException e) {
+            if(e.getCause() instanceof CodecNotFoundException)
+                throw (CodecNotFoundException)e.getCause();
+            throw new CodecNotFoundException(e.getCause(), cqlType, javaType);
+        } catch (ExecutionException e) {
+            throw new CodecNotFoundException(e.getCause(), cqlType, javaType);
+        }
+    }
+
+    private <T> TypeCodec<T> findCodec(DataType cqlType, TypeToken<T> javaType) {
+        checkNotNull(cqlType, "Parameter cqlType cannot be null");
+        logger.debug("Looking for codec [{} <-> {}]", cqlType, javaType == null ? "ANY" : javaType);
+        for (TypeCodec<?> codec : codecs) {
+            if (codec.accepts(cqlType) && (javaType == null || codec.accepts(javaType))) {
+                logger.debug("Codec found: {}", codec);
+                return (TypeCodec<T>)codec;
+            }
+        }
+        // codec not found among the reservoir of known codecs, try to create an ad-hoc codec;
+        // this situation should happen mainly when dealing with enums, tuples and UDTs,
+        // and collections thereof.
+        TypeCodec<T> codec = maybeCreateCodec(cqlType, javaType);
+        if(codec == null)
+            throw logErrorAndThrow(cqlType, javaType);
+        if (!codec.accepts(cqlType) || (javaType != null && !codec.accepts(javaType)))
+            throw logErrorAndThrow(cqlType, javaType);
+        logger.debug("Codec found: {}", codec);
+        codecs.addIfAbsent(codec);
+        return codec;
+    }
+
+    private <T> TypeCodec<T> findCodec(DataType cqlType, T value) {
```

Comment:
I agree that there is some duplication but I could not come up with a simple way to solve it, mainly because the code is not 100% identical, there are some slight changes in each variant (cf. the if block predicate).